src/provision/freedom-module.js

```diff
@@ -0,0 +1,351 @@
+    if (body !== null && typeof body !== "undefined") {
+      xhr.send({ string: body })
+    } else {
+      xhr.send(null);
+    }
+  }.bind(this));
+};
+
+/** 
+ * Waits for all in-progress Digital Ocean actions to complete
+ * e.g. after powering on a machine, or creating a VM
+ * @param {Function} resolve - call when done
+ * @param {Function} reject - call on failure
+ **/
+Provisioner.prototype._waitDigitalOceanActions = function(resolve, reject) {
+  console.log("Polling for Digital Ocean in-progress actions");
+  this._doRequest("GET", "droplets/" + this.state.cloud.vm.id + "/actions").then(function(resp) {
+    for (var i = 0; i < resp.actions.length; i++) {
+      if (resp.actions[i].status === "in-progress") {
+        setTimeout(this._waitDigitalOceanActions.bind(this, resolve, reject), POLL_TIMEOUT);
+        return;
+      }
+    }
+    resolve(resp);
+  }.bind(this)).catch(function(e) {
+    console.error("Error waiting for digital ocean actions:" + JSON.stringify(e));
+    reject(e)
+  }.bind(this));
+  
+};
+
+/**
+ * Properly configure Digital Ocean with a single droplet of name:name
+ * Assumes we already have oAuth token and  SSH key in this.state
+ * This method will use this._waitDigitalOceanActions() to wait until all actions complete
+ * before resolving
+ * @param {String} name of droplet
+ * @return {Promise} resolves on success, rejects on failure
+ **/
+Provisioner.prototype._setupDigitalOcean = function(name) {
+  return new Promise(function(resolve, reject) {
+    this.state.cloud = {};
+
+    this._sendStatus("CLOUD_INIT_ADDKEY");
+    // Get SSH keys in account
+    this._doRequest("GET", "account/keys").then(function(resp) {
+      //console.log(resp);
+      for (var i = 0; i < resp.ssh_keys.length; i++) {
+        if (resp.ssh_keys[i].public_key === this.state.ssh.public) {
+          return Promise.resolve({
+            message: "SSH Key is already in use on your account",
+            ssh_key: resp.ssh_keys[i]
+          });
+        } 
+      }
+      return this._doRequest("POST", "account/keys", JSON.stringify({
+        name: name,
+        public_key: this.state.ssh.public
+      }));
+    // If missing, put SSH key into account
+    }.bind(this)).then(function(resp) {
+      //console.log(resp);
+      this.state.cloud.ssh = resp.ssh_key;
+      this._sendStatus("CLOUD_DONE_ADDKEY");
+      this._sendStatus("CLOUD_INIT_VM");
+      return this._doRequest("GET", "droplets");
+    // Get list of droplets
+    }.bind(this)).then(function(resp) {
+      //console.log(resp);
+      for (var i = 0; i < resp.droplets.length; i++) {
+        if (resp.droplets[i].name === name) {
+          return Promise.resolve({
+            message: "Droplet already created with name=" + name,
+            droplet: resp.droplets[i]
+          });
+        }
+      }
+
+      return this._doRequest("POST", "droplets", JSON.stringify({
+        name: name,
+        region: "nyc3",
```

Comment:
For sure. I see the two key parameters being region and size. But we can tackle that in a followup.